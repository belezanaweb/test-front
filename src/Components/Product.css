.product-base {
  color: hsl(0, 0%, 0%);
  border: 1px hsl(0, 0%, 66%) solid;
  border-radius: 3px;
  margin: 5px;
  font-size: calc(10px + 1.1vmin);
}
