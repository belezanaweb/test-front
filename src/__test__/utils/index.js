import checkPropTypes from 'check-prop-types'

export const findByTestAtrr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`)
  return wrapper
}

export const checkProps = (component, expectedProps) => {
  const propsError = checkPropTypes(component.propTypes, expectedProps, 'props', component.name)
  return propsError
}
