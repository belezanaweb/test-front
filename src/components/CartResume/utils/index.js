export const normalizeCartData = data => {
  return {
    productsQuantity: data.items.length,
    subTotal: data.subTotal.toFixed(2).replace('.', ','),
    shipping: data.shippingTotal.toFixed(2).replace('.', ','),
    discount: data.discount.toFixed(2).replace('.', ','),
    total: data.total.toFixed(2).replace('.', ','),
  };
};
