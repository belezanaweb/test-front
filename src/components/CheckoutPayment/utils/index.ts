import * as Yup from 'yup'

const errors = {
  name: 'Insira um nome válido',
  code: 'Insira um código válido',
  number: 'Insira um número válido',
  expiration: 'Insira uma data válida',
}

export const initialValues = {
  creditCardName: '',
  creditCardNumber: '',
  creditCardExpiration: '',
  creditCardCode: '',
}

const regexOnlyNumbers = /[^\d.-]+/g

export const CheckoutPaymentSchema = Yup.object().shape({
  creditCardName: Yup.string()
    .min(2, errors.name)
    .matches(/^([A-Za-z\u00C0-\u00D6\u00D8-\u00f6\u00f8-\u00ff\s]*)$/gi, errors.name)
    .required(errors.name),
  creditCardNumber: Yup.string()
    .min(2, errors.number)
    .test('len', errors.number, (val) => {
      const value = val.replace(regexOnlyNumbers, '')

      return value.length === 16
    })
    .required(errors.number),
  creditCardExpiration: Yup.string()
    .min(2, errors.name)
    .test('len', errors.expiration, (val) => {
      const value = val.replace(regexOnlyNumbers, '')

      return value.length === 4
    })
    .required(errors.expiration),
  creditCardCode: Yup.string()
    .matches(/^[0-9]+$/, errors.code)
    .test('len', errors.code, (val) => val.length === 3)
    .required(errors.code),
})
