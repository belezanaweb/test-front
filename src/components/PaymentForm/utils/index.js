export const formatCardNumber = value => {
  return value
    .replace(/\D/g, '')
    .replace(/(\d{4})(?=\d)/g, '$1 ')
    .substr(0, 19);
};

export const formatHolderName = value => {
  return value.replace(/[^a-zA-Z\u00C0-\u017F\s]/gu, '').substr(0, 60);
};

export const formatExpirationDate = value => {
  return value
    .replace(/\D/g, '')
    .replace(/(\d{2})(\d{0,4})/, '$1/$2')
    .substr(0, 7);
};

export const formatCVVCode = value => {
  return value.replace(/\D/g, '').substr(0, 3);
};

const validateCreditCard = cardNumber => {
  cardNumber = cardNumber.replace(/\s/g, '');

  if (cardNumber.length !== 16) {
    return false;
  }

  const digits = cardNumber.split('').map(Number);

  // Algoritmo de Luhn para validar o número do cartão
  let sum = 0;

  for (let i = 0; i < digits.length; i++) {
    let digit = digits[i];

    if (i % 2 === 0) {
      digit *= 2;

      if (digit > 9) {
        digit -= 9;
      }
    }

    sum += digit;
  }

  // O número do cartão é válido se a soma dos dígitos for um múltiplo de 10
  return sum % 10 === 0;
};

const isInvalidDate = dateString => {
  const today = new Date();
  const [month, year] = dateString.split('/');
  if (month === '00') {
    return true;
  }
  const userDate = new Date(Number(year), Number(month) - 1);

  return userDate < new Date(today.getFullYear(), today.getMonth());
};

export const validateForm = state => {
  const errors = {};

  if (!validateCreditCard(state.cardNumber)) {
    errors.cardNumber = 'insira um número de cartão válido';
  }

  if (!state.cardHolderName) {
    errors.cardHolderName = 'insira um nome válido';
  }

  if (state.expirationDate.length < 7 || isInvalidDate(state.expirationDate)) {
    errors.expirationDate = 'insira uma data válida';
  }

  if (state.cvv.length < 3) {
    errors.cvv = 'insira um cvv válido';
  }

  return errors;
};
