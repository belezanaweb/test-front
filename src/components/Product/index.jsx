import React from 'react'
import { Container, ImageContainer, Image, TextContainer, ProductTitle, Price } from './styles'

export const Product = ({ name, priceSpecification, imageObjects }) => {
  return (
    <Container isConfirmation={window.location.pathname === '/confirmation'}>
      <ImageContainer>
        <img src={imageObjects[0]?.thumbnail} alt={name} />
      </ImageContainer>
      <TextContainer>
        <ProductTitle>{name}</ProductTitle>
        <Price isConfirmation={window.location.pathname === '/confirmation'}>
          R$ {priceSpecification.price.toFixed(2)}
        </Price>
      </TextContainer>
    </Container>
  )
}
