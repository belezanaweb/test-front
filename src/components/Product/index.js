import Product from './Product';

export default Product;
