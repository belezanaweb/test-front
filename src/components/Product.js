import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { formatToReal } from '../utils/currencyFormat'

const ProductImg = styled.div`
  margin-right: 1rem;
`

const ProductItem = styled.li`
  display: flex;
  padding: 1rem;
  margin: 1rem;
  border: var(--border-box);
`

const ProductDetails = styled.div`
  display:flex;
  flex: 1 1 auto;
  flex-direction: column;
  justify-content: space-between;
  font-size: 1.4rem;

  span:last-child {
    font-weight: bolder;
    align-self: flex-end;
  }
`

const Product = ({
  thumbnail,
  name,
  price
}) => (
  <ProductItem>
    <ProductImg>
      <img alt='Product' src={thumbnail} />
    </ProductImg>
    <ProductDetails>
      <span>{name}</span>
      <span>{formatToReal(price)}</span>
    </ProductDetails>
  </ProductItem>
);

Product.propTypes = {
  thumbnail: PropTypes.string,
  name: PropTypes.string,
  price: PropTypes.number
}

export default Product
