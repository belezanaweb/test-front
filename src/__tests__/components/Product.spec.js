import Product from '~/components/Product';

import { formatToCurrency } from '~/utils/format-currency';

const products = [
  {
    title: 'product 1',
    image: 'image 1',
    price: 10
  },
  {
    title: 'product 2',
    image: 'image 2',
    price: 20
  },
  {
    title: 'product 3',
    image: 'image 3',
    price: 30
  }
];

const [firstProduct] = products;

jest.mock('~/utils/format-currency', () => ({
  formatToCurrency: jest.fn()
}));

describe('Product Component', () => {
  it('should match a snapshot', () => {
    const wrapper = getStyledWrapperJSON(Product, {
      props: {
        ...firstProduct
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should call `formatToCurrency` to format price', () => {
    getStyledWrapper(Product, {
      props: {
        ...firstProduct
      }
    });

    expect(formatToCurrency).toHaveBeenCalled();
  });
});
