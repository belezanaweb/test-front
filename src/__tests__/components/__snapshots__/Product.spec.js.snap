// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component should match a snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border: 1px solid #ddd;
  border-radius: 3px;
  padding: 0.625rem;
  margin-bottom: 1rem;
}

.c0:last-child {
  margin-bottom: 0;
}

.c1 {
  -webkit-flex: 0 0 65px;
  -ms-flex: 0 0 65px;
  flex: 0 0 65px;
  width: 65px;
  margin-right: 0.625rem;
  margin-top: auto;
}

.c3 {
  font-weight: 400;
  font-size: 0.844rem;
  margin: 0 0 1.375rem;
}

.c4 {
  text-align: right;
  font-size: 0.875rem;
  font-weight: 700;
  display: block;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:768px) {
  .c0 {
    width: 33%;
    margin-right: 0.75rem;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
    padding: 1rem;
  }

  .c0,
  .c0:last-child {
    margin-bottom: 0;
  }

  .c0:last-child {
    margin-right: 0;
  }
}

@media (min-width:768px) {
  .c1 {
    margin: 0 auto auto;
    -webkit-flex: 0 0 130px;
    -ms-flex: 0 0 130px;
    flex: 0 0 130px;
    width: 130px;
  }
}

@media (min-width:768px) {
  .c4 {
    text-align: left;
  }
}

@media (min-width:768px) {
  .c2 {
    margin-top: 1rem;
  }
}

<div
  className="c0"
>
  <img
    className="c1"
    src="image 1"
  />
  <div
    className="c2"
  >
    <h3
      className="c3"
    >
      product 1
    </h3>
    <span
      className="c4"
    />
  </div>
</div>
`;
